File: leetcode/LinkedListTests/LinkedList_143.cs
namespace LinkedListTests;

[TestFixture]
class LinkedList_143
{
    class Solution {
        public void ReorderList(ListNode head) {
            var middle = GetMiddle(head);
            var reverseHead = Reverse(middle.next);
            middle.next = null;
            AdjacentMerge(head, reverseHead);

            void AdjacentMerge(ListNode head1, ListNode head2) {
                var curr1 = head1;
                var curr2 = head2;
            
                while(curr1 is not null && curr2 is not null) {
                    var next1 = curr1.next;
                    var next2 = curr2.next;
                    curr1.next = curr2;
                    curr2.next = next1;
                    curr1 = next1;
                    curr2 = next2;
                }
            }

            ListNode Reverse(ListNode node) {
                ListNode prev = null;
                var curr = node;
                while(curr is not null) {
                    var tmpNext = curr.next;
                    curr.next = prev;
                    prev = curr;
                    curr = tmpNext;
                }
                return prev;
            }

            ListNode GetMiddle(ListNode node) {
                var slow = node;
                var fast = node;
                while (fast is not null && fast.next is not null)
                {
                    fast = fast.next.next;
                    slow = slow.next;
                }
                return slow;
            }
        }
    }
}

File: leetcodePython/LinkedList/linkedlist_143.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    @staticmethod
    def _get_middle(head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        return slow

    @staticmethod
    def _reverse(head: Optional[ListNode]) -> Optional[ListNode]:
        previous = None
        current = head
        while current:
            current.next, previous, current = previous, current, current.next
        return previous

    def _adjacent_merge(self, head1: Optional[ListNode], head2: Optional[ListNode]):
        while head1 and head2:
            next1 = head1.next
            next2 = head2.next
            # head1 -> head2 ->
            head1.next = head2
            head2.next = next1
            head1 = next1
            head2 = next2

    def reorderList(self, head: Optional[ListNode]) -> None:
        middle = self._get_middle(head)
        reverse_middle = self._reverse(middle.next)
        middle.next = None
        self._adjacent_merge(head, reverse_middle)
